### PR TITLE
gh-107901: synthetic jumps which are not at end of loop no longer check the eval breaker

### PR DESCRIPTION
**Pull request by iritkatriel**
This PR changes some of the synthetic non-loop JUMPs to be JUMP_NO_INTERRUPT. 

The has two advantages:
1. These jumps have no line number, so we don't want to perform an eval breaker check in them.
2. It makes the code a bit smaller because there are no caches for these jumps.

* Issue: gh-107901

### CODE REVIEW

File: Python/flowgraph.c [hunk 1]
```diff
@@ -1602,6 +1602,7 @@ optimize_basic_block(PyObject *const_cache, basicblock *bb, PyObject *consts)
                 }
                 break;
             case JUMP:
+            case JUMP_NO_INTERRUPT:
```
Comment (markshannon) [inline Python/flowgraph.c:1613]: What about the opposite case, `JUMP` jumping to `JUMP_NO_INTERRUPT?
Now that `JUMP_NO_INTERRUPT` is more common, we should handle that case as well.

**Review by markshannon — CHANGES_REQUESTED** [review]
I think we need to convert `JUMP -> JUMP_NO_INTERRUPT` to `JUMP`.
Otherwise, looks good.

File: Python/flowgraph.c (hunk 1 above)
Comment (iritkatriel) [reply Python/flowgraph.c:1613]: I considered this, but the new JUMP_NO_INTERRUPT have no line number, so the current threading logic wouldn't do anything (unless the following jump also has no line number, which is unlikely). So the jump_thread function needs to change as well. I can do it in the same PR, if you don't think it complicates the review.

Comment (markshannon) [reply Python/flowgraph.c:1613]: It should be a fairly straightforward simple change to `jump_thread`.
If the initial jump has no line number, use the line number of the target.

Might as well do it in this PR.

Comment (iritkatriel) [reply Python/flowgraph.c:1613]: Once I add that I get assertion failed about some redundant jump not being cleared up, so I think that exposes another problem.

**Comment by iritkatriel** [comment]
I have made the requested changes; please review again.

File: Python/flowgraph.c (hunk 1 above)
Comment (iritkatriel) [reply Python/flowgraph.c:1613]: The problem was that the assertion in no_redundant_jumps wasn't checking line numbers, so it tripped when it shouldn't have. All good now.

**Comment by iritkatriel** [comment]
I have made the requested changes; please review again.

File: Python/flowgraph.c [hunk 2]
```diff
@@ -481,9 +481,12 @@ no_redundant_jumps(cfg_builder *g) {
             if (IS_UNCONDITIONAL_JUMP_OPCODE(last->i_opcode)) {
                 basicblock *next = next_nonempty_block(b->b_next);
                 basicblock *jump_target = next_nonempty_block(last->i_target);
-                assert(jump_target != next);
                 if (jump_target == next) {
-                    return false;
+                    assert(next);
+                    if (last->i_loc.lineno == next->b_instr[0].i_loc.lineno) {
+                        assert(0);
```
Comment (chris-eibl) [inline Python/flowgraph.c:487]: Is this `assert(0)` a leftover debugging help? Shouldn't it be removed before the final merge?

Comment (chris-eibl) [reply Python/flowgraph.c:487]: Sorry for the noise. I've just realized the old `assert(jump_target != next);` so presumably this should stay?

Comment (iritkatriel) [reply Python/flowgraph.c:487]: Yes, this function is only for debugging and it's nice if it crashes where the problem is.

**Review by markshannon — APPROVED** [review]
Looks good

**Pull request merged**